### PR TITLE
New client config setting: We can throw on every response other than 2xx and 304 responses if throwOnBadResponse is set in ClientConfig.

### PR DESCRIPTION
**Pull request by tarcang**
Note: With the current changes, the throwing will not happen if rawResponse is true.

### CODE REVIEW

File: src/test/crossFetchNode.test.ts [hunk 1]
```diff
@@ -274,3 +274,58 @@ test('should not fail when arbitrary parameters are configured in fetchOptions',
   });
   expect(response).toEqual({});
 });
+
+test('throws on responses other than and 304 errors', async () => {
```
Comment (wjhsf) [inline src/test/crossFetchNode.test.ts:278]:
```suggestion
test('throws on error responses', async () => {
```

File: src/test/crossFetchNode.test.ts [hunk 2]
```diff
@@ -274,3 +274,58 @@ test('should not fail when arbitrary parameters are configured in fetchOptions',
...
+    .matchHeader('authorization', 'Bearer test-auth')
+    .reply(400, {}, {'content-type': 'application-json charset=UTF-8'});
+
+  const standardConfig: ClientConfigInit<TestConfigParameters> = {...config};
+  const client = new ShopperSearch(standardConfig);
+  const response = async () =>
+    client.productSearch({
+      parameters: {q: 'sony'},
+      headers: {authorization: 'Bearer test-auth'},
+    });
```
Comment (wjhsf) [inline src/test/crossFetchNode.test.ts:296]:
```suggestion
  const responsePromise = client.productSearch({
      parameters: {q: 'sony'},
      headers: {authorization: 'Bearer test-auth'},
    });
```
The API call is an async function, so even though you added a `throw` statement, it doesn't throw. It returns a promise that gets rejected. You don't need to wrap the call to test it.

File: src/test/crossFetchNode.test.ts [hunk 3]
```diff
@@ -274,3 +274,110 @@ test('should not fail when arbitrary parameters are configured in fetchOptions',
...
+
+  expect(response).rejects.toThrow();
```
Comment (wjhsf) [inline src/test/crossFetchNode.test.ts:298]:
```suggestion
  await expect(response).rejects.toEqual(something);
```
1. `expect(promise).rejects` and `expect(promise).resolves` return a Promise, so you must `await` it, otherwise the checks are never actually done.
2. `expect(promise).rejects` basically unwraps the rejected error so that you can make assertions about it. You should make some assertions about it.

File: src/test/crossFetchNode.test.ts [hunk 4]
```diff
@@ -274,3 +274,110 @@ test('should not fail when arbitrary parameters are configured in fetchOptions',
...
+    )
+    .matchHeader('authorization', 'Bearer test-auth')
+    .reply(
+      400,
+      {content: 'not empty'},
+      {'content-type': 'application-json charset=UTF-8'}
+    );
+
+  const standardConfig: ClientConfigInit<TestConfigParameters> = {...config};
+  const client = new ShopperSearch(standardConfig);
```
Comment (wjhsf) [inline src/test/crossFetchNode.test.ts:291]:
```suggestion
  const client = new ShopperSearch({...config});
```
If you're not doing anything with it, you can just pass it directly. 😉

File: src/test/crossFetchNode.test.ts [hunk 5]
```diff
@@ -274,3 +274,110 @@ test('should not fail when arbitrary parameters are configured in fetchOptions',
...
+  const response = await client.productSearch({
+    parameters: {q: 'sony'},
+    headers: {authorization: 'Bearer test-auth'},
+  });
+
+  expect(response).toEqual({content: 'not empty'});
+});
+
+test('do not throw in empty json body', async () => {
```
Comment (wjhsf) [inline src/test/crossFetchNode.test.ts:367]: Also add a test case for an empty body.

**Review by wjhsf — CHANGES_REQUESTED** [review]

File: src/test/crossFetchNode.test.ts [hunk 6]
```diff
@@ -274,3 +274,105 @@ test('should not fail when arbitrary parameters are configured in fetchOptions',
...
+    .matchHeader('authorization', 'Bearer test-auth')
+    .reply(
+      400,
+      {content: 'not empty'},
+      {'content-type': 'application-json charset=UTF-8'}
+    );
+
+  const client = new ShopperSearch({...config});
+  const responsePromise = () =>
+    client.productSearch({
```
Comment (wjhsf) [inline src/test/crossFetchNode.test.ts:292]:
```suggestion
  const responsePromise = client.productSearch({
```
Functions aren't promises. 😉

File: src/test/crossFetchNode.test.ts [hunk 7]
```diff
@@ -274,3 +274,105 @@ test('should not fail when arbitrary parameters are configured in fetchOptions',
...
+      parameters: {q: 'sony'},
+      headers: {authorization: 'Bearer test-auth'},
+    });
+
+  expect(responsePromise).rejects.toEqual(new Error('Error 400: Bad Request'));
```
Comment (wjhsf) [inline src/test/crossFetchNode.test.ts:297]:
```suggestion
  await expect(responsePromise).rejects.toEqual(new Error('Error 400: Bad Request'));
```
Don't forget to await the expression so that it does what you expect it to!

**Review by wjhsf — DISMISSED** [review]

**Review by wjhsf — APPROVED** [review]

File: templates/operations.ts.hbs [hunk 8]
```diff
@@ -180,8 +180,13 @@
       const response = await fetch(req);
       if (rawResponse) {
         return response;
-      } else {
-        return response.json();
+      } else if (response.ok || response.status === 304) {
+        const text = await response.text();
+        // It's ideal to get "{}" for an empty response body, but we won't throw if it's truly empty
+        return text ? JSON.parse(text) : {};
```
Comment (wjhsf) [inline templates/operations.ts.hbs:186]: Actually, this may not be what we want to do here. Not all methods necessarily return a value.

**Review by wjhsf — CHANGES_REQUESTED** [review]

File: templates/operations.ts.hbs (hunk 8 above)
Comment (tarcang) [reply templates/operations.ts.hbs:186]: Which methods should I not return a value? Is there a doc that describes that?

File: templates/operations.ts.hbs [hunk 9]
```diff
@@ -180,8 +180,15 @@
       const response = await fetch(req);
       if (rawResponse) {
         return response;
-      } else {
-        return response.json();
+      } else if (response.ok || response.status === 304) {
+        {{#unless (eq (getReturnTypeFromOperation this) "void") }}
```
Comment (wjhsf) [inline templates/operations.ts.hbs:184]: This if block will be empty when the return type is `void`. I don't think the linter will like that.

Comment (tarcang) [reply templates/operations.ts.hbs:184]: I guess `return; ` is fine?

File: src/test/crossFetchNode.test.ts [hunk 10]
```diff
@@ -273,3 +273,131 @@ test('should not fail when arbitrary parameters are configured in fetchOptions',
...
+    headers: {
+      authorization: 'Bearer test-auth',
+    },
+    body: {
+      resetToken: 'R1e2s3e4t5T6o7k8e9n0',
+      login: 'janedoe@test.com',
+      newPassword: 'p@assword2',
+    },
+  });
+  expect(response);
```
Comment (wjhsf) [inline src/test/crossFetchNode.test.ts:402]:
```suggestion
  expect(response).toBeUndefined();
```
(Might not be the actual assertion, but I think it is...)

**Review by wjhsf — APPROVED** [review]

File: src/test/crossFetchNode.test.ts (hunk 10 above)
Comment (tarcang) [reply src/test/crossFetchNode.test.ts:402]: 👌

File: templates/operations.ts.hbs [hunk 11]
```diff
@@ -180,8 +180,17 @@
...
+      } else if (response.ok || response.status === 304) {
+        {{#if (eq (getReturnTypeFromOperation this) "void") }}
+          return;
+        {{else}}
+          const text = await response.text();
+          // It's ideal to get "{}" for an empty response body, but we won't throw if it's truly empty
+          return text ? JSON.parse(text) : {};
+        {{/if}}
+      }
+      else {
```
[the remark this answers is not in the document]
Comment (tarcang) [reply templates/operations.ts.hbs:192]: 👌

File: src/test/crossFetchNode.test.ts (hunk 10 above)
Comment (tarcang) [reply src/test/crossFetchNode.test.ts]: It is the assertion :)

**Review by wjhsf — APPROVED** [review]

File: templates/operations.ts.hbs [hunk 12]
```diff
@@ -115,6 +117,7 @@
         body: {{{getPayloadTypeFromRequest request}}}{{/or}}
       }>,
       rawResponse?: boolean,
+      throwOnBadResponse?: boolean
```
Comment (wjhsf) [inline templates/operations.ts.hbs:120]: Throwing on bad responses is probably a behavior that the SDK user wants to be consistent every time.  This should be set on the client config, rather than being passed every single time they call a method.

File: templates/operations.ts.hbs [hunk 13]
```diff
@@ -180,10 +183,26 @@
...
+                throw new Error('Error ' + response.status + ': ' + response.statusText);
+              }
       } else {
-        return response.json();
+        //If rawResponse and throwOnBadResponse is false, go with default behavior.
+        {{#if (eq (getReturnTypeFromOperation this) "void") }}
+          return;
+        {{else}}
+          return response.json();
+        {{/if}}
```
Comment (wjhsf) [inline templates/operations.ts.hbs:204]: You can flatten this logic a bit.
```ts
if (rawResponse) {
  return response;
} else if (throwOnBadResponse && !response.ok && response.status !== 304) {
  throw;
} else {
  /* void or JSON block */
}
```

File: src/test/crossFetchNode.test.ts [hunk 14]
```diff
@@ -273,3 +273,201 @@ test('should not fail when arbitrary parameters are configured in fetchOptions',
...
+    );
+
+  const client = new ShopperSearch({...config});
+  const response = await client.productSearch({
+    parameters: {q: 'sony'},
+    headers: {authorization: 'Bearer test-auth'},
+  });
+
+  expect(response).toEqual({content: 'not empty'});
+});
```
Comment (wjhsf) [inline src/test/crossFetchNode.test.ts:473]: Don't forget to add a test case for `true`!

**Review by wjhsf — CHANGES_REQUESTED** [review]

File: src/static/clientConfig.ts [hunk 15]
```diff
@@ -53,6 +54,8 @@ export default class ClientConfig<Params extends BaseUriParameters>
     ClientConfigInit<Params>['transformRequest']
   >;
 
+  public throwOnBadResponse?: boolean = false;
```
Comment (wjhsf) [inline src/static/clientConfig.ts:57]:
```suggestion
  public throwOnBadResponse: boolean;
```
Makes life easier if you have fewer possible states to deal with. 😉 

(Should remain optional in `ClientConfigInit`.)

File: src/static/clientConfig.ts [hunk 16]
```diff
@@ -71,6 +74,9 @@ export default class ClientConfig<Params extends BaseUriParameters>
     if (config.proxy) {
       this.proxy = config.proxy;
     }
+    if (config.throwOnBadResponse !== undefined) {
+      this.throwOnBadResponse = config.throwOnBadResponse;
+    }
```
Comment (wjhsf) [inline src/static/clientConfig.ts:79]:
```suggestion
this.throwOnBadResponse = !!config.throwOnBadResponse;
```
Ensures that the property is always a boolean.

File: templates/operations.ts.hbs [hunk 17]
```diff
@@ -180,10 +179,18 @@
...
       } else {
-        return response.json();
+        //If rawResponse and throwOnBadResponse is false, go with default behavior.
+        {{#if (eq (getReturnTypeFromOperation this) "void") }}
+          return;
+        {{else}}
+          const text = await response.text();
+          // It's ideal to get "{}" for an empty response body, but we won't throw if it's truly empty
+          return text ? JSON.parse(text) : {};
+        {{/if}}
```
Comment (wjhsf) [inline templates/operations.ts.hbs:192]: After the refactor, this is the last block in the function. The return statement for `void` has become redundant, because there's literally nothing after it. The whole `else` code block can be moved into the non-void return type template block.

File: src/static/clientConfig.ts [hunk 18]
```diff
@@ -31,6 +31,7 @@ export interface ClientConfigInit<Params extends BaseUriParameters> {
     data: T,
     headers: {[key: string]: string}
   ) => Required<FetchOptions>['body'];
+  throwOnBadResponse?: boolean;
```
Comment (bredmond-sf) [inline src/static/clientConfig.ts:34]: There is a story to redo the README, but please add a description of this new parameter to to current one so it can easily be moved to the new one!

**Review by bredmond-sf — CHANGES_REQUESTED** [review]

File: templates/operations.ts.hbs [hunk 19]
```diff
@@ -180,10 +179,16 @@
       const response = await fetch(req);
       if (rawResponse) {
         return response;
+      } else if (this.clientConfig.throwOnBadResponse && !response.ok && response.status !== 304) {
+        throw new Error('Error ' + response.status + ': ' + response.statusText);
       } else {
-        return response.json();
+        //If rawResponse and throwOnBadResponse is false, go with default behavior.
```
Comment (wjhsf) [inline templates/operations.ts.hbs:185]: What _is_ the default behavior? Is this comment still accurate?

File: templates/operations.ts.hbs (hunk 9 above)
Comment (wjhsf) [reply templates/operations.ts.hbs:186]: The entire `else` block should be inside the `{{#unless}}`, otherwise you'll end up with
```ts
else { }
```

File: README.md [hunk 20]
```diff
@@ -90,6 +90,10 @@ const config = {
 }
  ```
 
+### Optional settings that can be set in the config object
+*throwOnBadResponse:* Default value is false. When set to true, the sdk with throw on exception on responses with statuses that is not
```
Comment (bredmond-sf) [inline README.md:94]:
```suggestion
*throwOnBadResponse:* Default value is false. When set to true, the sdk will throw an Error on responses with statuses that are not
```

**Review by bredmond-sf — DISMISSED** [review]

**Review by wjhsf — DISMISSED** [review]

**Review by joeluong-sfcc — DISMISSED** [review]

**Review by unandyala — DISMISSED** [review]

**Review by wjhsf — APPROVED** [review]

**Review by joeluong-sfcc — APPROVED** [review]
